tsweb/precompress.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsweb
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+	"io/fs"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/andybalholm/brotli"
+	"golang.org/x/sync/errgroup"
+)
+
+// PrecompressDir compresses static assets in dirPath using Gzip and Brotli, so
+// that they can be later served with OpenPrecompressedFile. Meant to be run
+// during build steps, to avoid the cost of repeatedly compressing unchanging
+// resources.
+func PrecompressDir(dirPath string, options PrecompressOptions) error {
+	var eg errgroup.Group
+	err := fs.WalkDir(os.DirFS(dirPath), ".", func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		if !compressibleExtensions[filepath.Ext(p)] {
+			return nil
+		}
+		p = path.Join(dirPath, p)
+		if options.ProgressFn != nil {
+			options.ProgressFn(p)
+		}
+
+		eg.Go(func() error {
+			return precompress(p, options)
+		})
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return eg.Wait()
+}
+
+type PrecompressOptions struct {
+	// FastCompression controls whether compression should be optimized for
+	// speed rather than size.
+	FastCompression bool
+	// ProgressFn is invoked when a file in the directory is about to be
+	// compressed.
+	ProgressFn func(path string)
```

Comment:
is this optional?

Reply:
Yes. Added a comment, is there a canonical way to indicate that it is?

Reply:
Looks like a "if non-nil" clause is idiomatic.